### DISCUSSION TITLE
Carrier Billing Certification

### DISCUSSION THREAD

**Discussion by ajmhussain** [camaraproject/CarrierBillingCheckOut#198] [Q&A]
Dear Team (@PedroDiez , @diegogonmar )

We have implemented carrier billing v0.2.1 but no test definitions exist for that, and we are now planning to upgrade to version 0.3.x. However, from certification perspective we have few queries which need answers from GSMA/CAMARA:
1. We haven't implemented callbacks as those are not applicable in our use-cases. Can we certify without implementing callbacks?
2. We haven't implemented payment validation API again for the same reason-- not in our business requirements scope. Can we get certified without implementing validate use-case?

Please respond to the above queries so that we plan accordingly. 

Regards
Ajmal Hussain
Etisalat UAE

**Comment by PedroDiez** [comment]
Hi @ajmhussain,

The first version of Carrier Billing aligned with CAMARA MetaRelease Tracks (first MetaRelease has been Fall 24), has been v0.3.x. Following CAMARA Release Management guidelines, the first Test Definitions has been defined for v0.3.x (there will not be Test Definitions for v0.2.1).

Test Definitions for v0.3.x: https://github.com/camaraproject/CarrierBillingCheckOut/tree/r1.3/code/Test_definitions

1. We haven't implemented callbacks as those are not applicable in our use-cases. Can we certify without implementing callbacks?

> Not mandatory to implement callbacks, specifically if Telco Operator Behaviour is sync. This is up to Business and also related to integration agreements for the Operator. So as, honestly, this should not be a matter. Just in case a Telco Operator needs this in their business later, rational way would be to notify GSMA and certify the behaviour supporting callbacks

2. We haven't implemented payment validation API again for the same reason-- not in our business requirements scope. Can we get certified without implementing validate use-case?

> Not mandatory to implement "payment validation" feature, specifically if Telco Operator does not need such a future in their business cases. It is an optional feature that may apply in a 2-step payment scenario only. So, the same approach applies, this should not be a matter. Just in case a Telco Operator needs this in their business later, rational way would be to notify GSMA and certify the behaviour supporting "payment validation"

**Comment by ajmhussain** [comment]
Thanks @PedroDiez for your prompt response. 
A few additional queries requiring your input:

1.  We have only considered Carrier Billing (Payment) API without refund API (which is a separate swagger). For certification, will it be good to go? I believe Refund API can be certified separately later?
2. In-case of business failure from billing system kindly confirm if message text is flexible as it can vary according to use-case and may not be limited to only messages mentioned in camara docs?
3. To meet our business use-cases, For the following APIs can we define a new additional header parameter (_merchantIdentifier_) so that a client can inquire/confirm/cancel payments related to that specific merchant only:

- GET payments
- POST Payments/{paymentId}/cancel
- POST Payments/{paymentId}/confirm

Appreciate your support on addressing the queries. 

Thanks & Regards
Ajmal Hussain
Etisalat AE

**Reply by PedroDiez** [reply]
Hi @ajmhussain 

1. We have only considered Carrier Billing (Payment) API without refund API (which is a separate swagger). For certification, will it be good to go? I believe Refund API can be certified separately later?

> It is absolutely possible. Yes, you are right, a Telco Operator can work/expose Carrier Billing (Payment) and not expose Carrier Billing Refund (i.e. it manages by other means in the emantime). Later, according to their needs/business roadmap, can incorporate Carrier Billing Refund and then certification for the API can be performed. 

2. In-case of business failure from billing system kindly confirm if message text is flexible as it can vary according to use-case and may not be limited to only messages mentioned in camara docs?

> Yes, confirmed. `message` field in model errors can be adapted to provide the information a Telco Operator consider of relevance un an error context. It is an `informative` field 

3. To meet our business use-cases, For the following APIs can we define a new additional header parameter (merchantIdentifier) so that a client can inquire/confirm/cancel payments related to that specific merchant only:
- GET payments

> It is already covered, GET /payments has a queryParam named `merchantIdentifier`

- POST Payments/{paymentId}/cancel
- POST Payments/{paymentId}/confirm

> This is a 2-STEP payment. The indication of `merchantIdentifier`, if required by business, has to be in the first step (a.k.a. preparePayment) POST /payments/prepare (Request Body: amountTransaction -> paymentAmount -> chargingMetaData -> merchantIdentifier). This process generates a `paymentId`. So cancel/confirm is referred to a `paymentId` and implicitly associated to that `merchantIdentifier`.


Best Regards
Pedro

**Comment by ajmhussain** [comment]
Hi @PedroDiez 
Your response on above additional queries is awaited. Appreciate your help.

Regards
Ajmal Hussain

**Comment by ajmhussain** [comment]
Dear @PedroDiez , @diegogonmar 

Can you please help providing test scripts for Carrier Billing Payment APIs certification so that we may run the test scripts and validate the APIs before formally requesting for certification. We can see test definitions in the repo but we require test scripts so that our testing team may execute those directly. We were given test scripts by your team in past when we went through similar certification exercise for SIM SWAP API. 

Appreciate your support and prompt responses.

Regards
Ajmal Hussain
Etisalat UAE

**Comment by ajmhussain** [comment]
@PedroDiez , @diegogonmar 

Happy new year. Please help providing postman collection of the test cases for carrier billing payment APIs so that we may run those internally as part of our preparation activities for certification. Thanks

Regards
Ajmal Hussain
Etisalat UAE

**Reply by PedroDiez** [reply]
Hi @ajmhussain,

Apologize for delay, we were on holidays. Happy new year to you as well.

Within CAMARA Subprojects it is not part of their work to provide test scripts and/or postman collections for the certification of a given API. AFAIK it is GSMA the one that consider gherkin testing in order to generate postman for certification.

Best Regards
Pedro

**Comment by ajmhussain** [comment]
Hi @PedroDiez 

Thanks for your response. For another API (Sim Swap) we already got certification and for that we got postman collection from your colleague Toyeeb Rehman <[trehman@gsma.com](mailto:trehman@gsma.com)>

We are expecting similar process for this API certification too hence requested for test scripts postman collection.  Please advise

Regards
Ajmal Hussain
Etisalat AE
